wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.wsplugins;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.util.Collections.emptyList;
+
+import com.google.common.annotations.Beta;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.wsplugins.model.PluginFQN;
+import org.eclipse.che.api.workspace.shared.Constants;
+
+/**
+ * Parses workspace attributes into a list of {@link PluginFQN}.
+ *
+ * <p>This API is in <b>Beta</b> and is subject to changes or removal.
+ *
+ * @author Oleksander Garagatyi
+ * @author Angel Misevski
+ */
+@Beta
+public class PluginFQNParser {
+
+  public Collection<PluginFQN> parsePlugins(Map<String, String> attributes)
+      throws InfrastructureException {
+    if (attributes == null) {
+      return emptyList();
+    }
+
+    String pluginsAttribute =
+        attributes.getOrDefault(Constants.WORKSPACE_TOOLING_PLUGINS_ATTRIBUTE, null);
+    String editorAttribute =
+        attributes.getOrDefault(Constants.WORKSPACE_TOOLING_EDITOR_ATTRIBUTE, null);
+
+    List<PluginFQN> metaFQNs = new ArrayList<>();
+    if (!isNullOrEmpty(pluginsAttribute)) {
+      String[] plugins = pluginsAttribute.split(" *, *");
```

Comment:
` *, *` makes sure that empty spaces will be removed between plugins but it does not remove empty spaces before the first plugin and after the last plugin. Maybe it would be safer to split by `,` and then trim split values.
```java
  @Test(
      expectedExceptions = InfrastructureException.class,
      expectedExceptionsMessageRegExp = "Invalid.*duplicated")
  public void shouldDetectDuplicatedPluginsIfTheyArePrefixedSuffixedWithEmptySpaces() throws Exception {
    Map<String, String> attributes =
        createAttributes(
            "",
            " " + formatPlugin("http://testregistry1:8080", "testplugin", "1.0"),
            formatPlugin("http://testregistry2:8080", "testplugin", "1.0") + " ");

    parser.parsePlugins(attributes);
  }
```

Reply:
Good catch -- done. And thanks for the test case, I've added it to the suite.